client/client.go

```diff
@@ -0,0 +1,421 @@
+package client
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/notary/trustmanager"
+	"github.com/endophage/gotuf"
+	"github.com/endophage/gotuf/data"
+	"github.com/endophage/gotuf/keys"
+	"github.com/endophage/gotuf/signed"
+	"github.com/endophage/gotuf/store"
+	"github.com/spf13/viper"
+)
+
+// Default paths should end with a '/' so directory creation works correctly
+const trustDir string = "/trusted_certificates/"
+const privDir string = "/private/"
+const tufDir string = "/tuf/"
+const rootKeysDir string = "/root_keys/"
+
+// ErrRepositoryNotExist gets returned when trying to make an action over a repository
+/// that doesn't exist
+var ErrRepositoryNotExist = errors.New("repository does not exist")
+
+// Client is the interface that defines the Notary Client type
+type Client interface {
+	ListPrivateKeys() []data.PrivateKey
+	GenRootKey(passphrase string) (*data.PublicKey, error)
+	GetRepository(gun string, baseURL string, transport http.RoundTripper) (Repository, error)
+}
+
+type NotaryClient struct {
+	configFile       string
+	caStore          trustmanager.X509Store
+	certificateStore trustmanager.X509Store
+	rootKeyStore     trustmanager.EncryptedFileStore
+}
+
+type NotaryRepository struct {
+	Gun          string
+	baseURL      string
+	transport    http.RoundTripper
+	signer       *signed.Signer
+	tufRepo      *tuf.TufRepo
+	fileStore    store.MetadataStore
+	privKeyStore trustmanager.FileStore
+}
+
+// Target represents a simplified version of the data TUF operates on.
+type Target struct {
+	Name   string
+	Hashes data.Hashes
+	Length int64
+}
+
+// NewTarget  is a helper method that returns a Target
+func NewTarget(targetName string, targetPath string) (*Target, error) {
+	b, err := ioutil.ReadFile(targetPath)
+	if err != nil {
+		return nil, err
+	}
+
+	meta, err := data.NewFileMeta(bytes.NewBuffer(b))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Target{Name: targetName, Hashes: meta.Hashes, Length: meta.Length}, nil
+}
+
+// Repository is the interface that represents a Notary Repository
+type Repository interface {
+	Update() error
+	Initialize(key *data.PublicKey) error
+
+	AddTarget(target *Target) error
+	ListTargets() ([]*Target, error)
+	GetTargetByName(name string) (*Target, error)
+
+	Publish() error
+}
+
+// NewClient is a helper method that returns a new notary Client, given a config
+// file. It makes the assumption that the base directory for the config file will
+// be the place where trust information is being cached locally.
+func NewClient(configFile string) (*NotaryClient, error) {
```

Comment:
This shouldn't be doing config parsing. Pass in the config values themselves. Both our commands and docker/docker already do config parsing.